gen/AArch64Compile_gen.ml

```diff
@@ -2158,6 +2175,20 @@ module Make(Cfg:Config) : XXXCompile_gen.S =
     | Cas -> map_some_dp emit_cas_dep
     | StOp op -> emit_stop_dep op
 
+    let emit_fence_dp st p init n f (dp,csel) r1 n1 =
+      let vdep = node2vdep n1 in
+      match dp with
+      | D.ADDR
+      | D.DATA -> let init,cs, st = emit_fence st p init n f in
```

Comment:
I tried generating some examples with DpAddr and DpCtrl but I didn't get what I was expecting. For example:

```
$> diyone7 -moreedges true -arch AArch64 -variant self "DMB.ISHdWW Rfe [DpAddrdR ISB DC.CVAUp DSB.ISH IC.IVAUp DSB.ISH ISB Ifre]"
```

Throws an error that there can't be an address dependency to code. Which makes sense. But then I think I don't know how to use DpAddr.

Reply:
Hmm, that is strange, what behaviour would you expect here? Ctrl does seem to work, but not Addr. should we expect both to work or neither?

Reply:
I am not sure what are the meaningful tests we can produce with DpAddr*R then. For example, if we ignore the address dependency to code location:

```
$> diyone7 -moreedges true -arch AArch64 -variant self "DMB.ISHdWW Rfe [DpAddrdR DC.CVAUn DSB.ISH IC.IVAUn DSB.ISH ISB] Fre"
AArch64 A
"DMB.ISHdWW Rfe DpAddrdR DC.CVAUn DSB.ISH IC.IVAUn DSB.ISH ISB Fre"
Generator=diyone7 (version 7.56+03)
Com=Rf Fr
Orig=DMB.ISHdWW Rfe DpAddrdR DC.CVAUn DSB.ISH IC.IVAUn DSB.ISH ISB Fre
{
0:X1=x; 0:X3=y;
1:X0=y; 1:X2=x;
}
 P0          | P1                  ;
 MOV W0,#1   | LDR W1,[X0]         ;
 STR W0,[X1] | DC CVAU,X2          ;
 DMB ISH     | DSB ISH             ;
 MOV W2,#1   | IC IVAU,X2          ;
 STR W2,[X3] | DSB ISH             ;
             | ISB                 ;
             | EOR W3,W1,W1        ;
             | LDR W4,[X2,W3,SXTW] ;
exists (1:X1=1 /\ 1:X4=0)
```

The DC doesn't have an address dependency from the 1st LDR in P1.

Reply:
Ignoring ifetch for a minute:
```
$> diyone7 -moreedges true -arch AArch64 -variant self "DMB.ISHdWW Rfe DpAddrdR DC.CVAUn DSB.ISH IC.IVAUn DSB.ISH ISB Fre"
AArch64 A
"DMB.ISHdWW Rfe DpAddrdR DC.CVAUn DSB.ISH IC.IVAUn DSB.ISH ISB Fre"
Generator=diyone7 (version 7.56+03)
Com=Rf Fr
Orig=DMB.ISHdWW Rfe DpAddrdR DC.CVAUn DSB.ISH IC.IVAUn DSB.ISH ISB Fre
{
0:X1=x; 0:X3=y;
1:X0=y; 1:X2=x;
}
 P0          | P1                  ;
 MOV W0,#1   | LDR W1,[X0]         ;
 STR W0,[X1] | DC CVAU,X2          ;
 DMB ISH     | DSB ISH             ;
 MOV W2,#1   | IC IVAU,X2          ;
 STR W2,[X3] | DSB ISH             ;
             | ISB                 ;
             | EOR W3,W1,W1        ;
             | LDR W4,[X2,W3,SXTW] ;
exists (1:X1=1 /\ 1:X4=0)
```

But I would expect:

```
AArch64 A
"DMB.ISHdWW Rfe DpAddrdR DC.CVAUn DSB.ISH IC.IVAUn DSB.ISH ISB Fre"
Generator=diyone7 (version 7.56+03)
Com=Rf Fr
Orig=DMB.ISHdWW Rfe DpAddrdR DC.CVAUn DSB.ISH IC.IVAUn DSB.ISH ISB Fre
{
0:X1=x; 0:X3=y;
1:X0=y; 1:X2=x;
}
 P0          | P1                  ;
 MOV W0,#1   | LDR W1,[X0]         ;
             | EOR W3,W1,W1        ;
             | ADD X5,X2,W3,SXTW   ;
 STR W0,[X1] | DC CVAU,X5          ;
 DMB ISH     | DSB ISH             ;
 MOV W2,#1   | IC IVAU,X5          ;
 STR W2,[X3] | DSB ISH             ;
             | ISB                 ;
             | LDR W4,[X5] ;
exists (1:X1=1 /\ 1:X4=0)
```

And if it won't be too hard, I think for ifetch we should get this test:
```
diyone7 -moreedges true -arch AArch64 -variant self "DMB.ISHdWW Rfe DpAddrdR DC.CVAUn DSB.ISH IC.IVAUn DSB.ISH ISB Ifre"
AArch64 A
"DMB.ISHdWW Rfe DpAddrdR DC.CVAUn DSB.ISH IC.IVAUn DSB.ISH ISB Ifre"
Generator=diyone7 (version 7.56+03)
Com=Rf Ifr
Orig=DMB.ISHdWW Rfe DpAddrdR DC.CVAUn DSB.ISH IC.IVAUn DSB.ISH ISB Ifre
{
0:X0=NOP; 0:X1=P1:Lself00; 0:X3=x;
1:X0=x; 1:X2=P1:Lself00;
}
 P0          | P1          ;
 STR W0,[X1] | LDR W1,[X0] ;
             | EOR W4,W1,W1 ;
             | ADD X3,X2,W4,SXTW ;
 DMB ISH     | DC CVAU,X3  ;
 MOV W2,#1   | DSB ISH     ;
 STR W2,[X3] | IC IVAU,X3  ;
             | DSB ISH     ;
             | ISB         ;
             | Lself00:    ;
             | B L00       ;
             | MOV W3,#2   ;
             | B L01       ;
             | L00:        ;
             | MOV W3,#1   ;
             | L01:        ;
exists (1:X1=1 /\ 1:X3=1)
```

Reply:
Hi Nikos, I have implemented this change now for the normal case without ifetch, however the case with ifetch is a littl more complicated so for now I am throwing an error. Let me know if it is now doing what you expect it to.

I will now also rebase the patch on top of master